src/extension.ts

```diff
@@ -181,10 +183,40 @@ export async function activate(context: ExtensionContext) {
     context.subscriptions
   );
 
+  const TextDocumentListener = workspace.onDidSaveTextDocument(
+    async document => {
+      const workspaceDenoConfig = workspace.getConfiguration(
+        configurationSection
+      );
+      const denoEnable = workspaceDenoConfig.get("enable", true);
+      const autoFmtOnSave = workspaceDenoConfig.get("autoFmtOnSave", false);
+
+      // if not enable the deno or not enable ``auto format on save``
+      if (!denoEnable || !autoFmtOnSave) {
+        return;
+      }
+
+      // TODO: ignore the large file. it will slowdown the extension
+      if (isTypeScriptDocument(document) || isJavaScriptDocument(document)) {
+        try {
+          // !!!: deno only support relative file path in currently
+          const filename = path.basename(document.uri.fsPath);
+          const cwd = path.dirname(document.uri.fsPath);
+          await execa("deno", [filename, "--fmt"], {
+            cwd
+          });
+        } catch {
+          // NOBUG: ignore error
+        }
+      }
+    }
+  );
+
```

Comment:
ahould we register `languages.registerDocumentFormattingEditProvider` and `languages.registerDocumentRangeFormattingEditProvider` instead of directly formatting it?

- https://code.visualstudio.com/api/references/vscode-api#languages

Reply:
yes. I will re-push soon

Reply:
`languages.registerDocumentFormattingEditProvider` is only for `formatter`.

If we want to implement automatic formatting when saving.

There are 2 methods:

1. listen on `workspace.onDidSaveTextDocument` and trigger `format`
2. set build-in setting `Format On Save` to `true` and set `deno` formatter to default. The bad is `Format On Save` effect all language